### PR TITLE
erc 721 coin instance page

### PR DESCRIPTION
**Pull request by ayrat555**
resolves https://github.com/poanetwork/blockscout/issues/2541

ERC721 token instance pages can be accessed from the token inventory tab

## Changelog
- erc 721 coin instance page

### CODE REVIEW

File: apps/block_scout_web/lib/block_scout_web/controllers/tokens/instance_controller.ex [hunk 1]
```diff
@@ -0,0 +1,21 @@
+defmodule BlockScoutWeb.Tokens.InstanceController do
+  use BlockScoutWeb, :controller
+
+  alias Explorer.Chain
+
+  def show(conn, %{"token_id" => token_id, "id" => token_address_hash}) do
+    with {:ok, hash} <- Chain.string_to_address_hash(token_address_hash),
+         {:ok, _token_address} <- Chain.token_from_address_hash(hash, []),
```
Comment (pasqu4le) [inline apps/block_scout_web/lib/block_scout_web/controllers/tokens/instance_controller.ex]: Given that this is only to check the token existence, this fits a case that has been found before and there are functions written for it specifically in `Explorer.Chain`.

I think you should add a `check_token_exists`\`token_exists?` and use those instead.
You can check one of the others for reference, they are all at the end of the module.

File: apps/block_scout_web/lib/block_scout_web/controllers/tokens/instance_controller.ex [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+         {:ok, _token_transfer} <-
+           Chain.erc721_token_instance_from_token_id_and_token_address(token_id, hash) do
```
Comment (pasqu4le) [inline apps/block_scout_web/lib/block_scout_web/controllers/tokens/instance_controller.ex]: Same as above, as it is just checking for existence

File: apps/block_scout_web/lib/block_scout_web/controllers/tokens/instance_controller.ex (hunk 1 above)
Comment (ayrat555) [reply apps/block_scout_web/lib/block_scout_web/controllers/tokens/instance_controller.ex]: good idea

File: apps/block_scout_web/lib/block_scout_web/controllers/tokens/instance_controller.ex (hunk 2 above)
Comment (ayrat555) [reply apps/block_scout_web/lib/block_scout_web/controllers/tokens/instance_controller.ex]: done

**Comment by vbaranov** [comment]
@ayrat555 in addition
- stuck tx tile token ticker - value in tx tile (see stg for examples). I don't see this bug in master branch.

Screenshot 2019-09-10 at 18 44 48

- https://...//tokens/10306360948590189/instance/0x5caebd3b32e210e85ce3e9d51638b9c445481567 - wrong link: doubled slash before `tokens`, `https` is instead of `http` for stg
- I suggest changing the token instance link to 
```
http(s)://.../tokens/{token_hash}/instance/{token_id}
```
token_hash -> token ID (analogy to class -> instance) instead of `/tokens/{token_id}/instance/{token_hash}`

File: CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,7 @@
 - [#2666](https://github.com/poanetwork/blockscout/pull/2666) - fetch token counters in parallel
 - [#2665](https://github.com/poanetwork/blockscout/pull/2665) - new menu layout for mobile devices
 - [#2663](https://github.com/poanetwork/blockscout/pull/2663) - Fetch address counters in parallel
+- [#2642](https://github.com/poanetwork/blockscout/pull/2642) - add ERC721 coin instance page
```
Comment (pasqu4le) [inline CHANGELOG.md]: Just a nitpick, but this should be on top of the list

File: apps/block_scout_web/lib/block_scout_web/templates/tokens/instance/metadata/index.html.eex [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+          <div class="d-flex justify-content-between align-items-baseline">
+            <h3><%= gettext "Metadata" %></h3>
+            <button type="button" class="btn-line" id="button" data-clipboard-text="<%= format_metadata(@token_instance.instance.metadata) %>" aria-label="Copy Metadata">
+              <%= gettext "Copy Metadata" %>
+            </button>
+          </div>
+          <div class="tile tile-muted mb-4">
+            <pre class="pre-wrap pre-scrollable"><code class="nohighlight"><%= format_metadata(@token_instance.instance.metadata) %></code>
+            </pre>
+                </div>
```
Comment (pasqu4le) [inline apps/block_scout_web/lib/block_scout_web/templates/tokens/instance/metadata/index.html.eex]: This `</div>` seems unnecessary

File: apps/explorer/lib/explorer/chain/token_transfer.ex [hunk 5]
```diff
@@ -140,6 +148,26 @@ defmodule Explorer.Chain.TokenTransfer do
...
   end
 
+  @spec fetch_token_transfers_from_token_hash_and_token_id(Hash.t(), binary(), [paging_options]) :: []
+  def fetch_token_transfers_from_token_hash_and_token_id(token_address_hash, token_id, options) do
+    paging_options = Keyword.get(options, :paging_options, @default_paging_options)
+
+    query =
+      from(
+        tt in TokenTransfer,
+        where:
```
Comment (pasqu4le) [inline apps/explorer/lib/explorer/chain/token_transfer.ex]: IMHO, just for readability, I'd suggest to split this in multiple `where`s:
```elixir
where: tt.token_contract_address_hash == ^token_address_hash,
where: tt.token_id == ^token_id,
where: not is_nil(tt.block_number),
```

**Comment by ayrat555** [comment]
@vbaranov I fixed it

**Comment by vbaranov** [comment]
@ayrat555 could you please fix merging conflicts?

**Pull request merged**